docs/testplan/Upgrade_gNOI-test-plan.md

```diff
@@ -0,0 +1,337 @@
+
+
+2) No management interface (e.g., eth0 down)
+- Setup: administratively disable DUT management interface.
+- Expected: control host cannot open gRPC connection; agent times out with clear connectivity error.
+- Verify: grpc connection attempts fail; test harness records timeout; recovery steps are documented.
+
+3) No route to firmware server
+- Setup: block firmware server via iptables or remove route on DUT/control host.
+- Expected: remote download fails with network error; agent retries according to policy and then errors out.
+- Verify: agent error codes indicate network/dns failure; download sessions are cleaned up.
+
+4) Corrupt or tampered image (checksum mismatch)
+- Setup: serve an image whose contents do not match provided SHA‑256 digest (or omit digest).
+- Expected: server or DUT verifies digest and rejects image; install not performed.
+- Verify: checksum mismatch logged; installation not started; proper error returned.
+
+5) TLS / certificate failures
+- Setup: present incorrect TLS certs or force TLS mismatch between agent and gNOI server.
+- Expected: TLS handshake fails and gRPC connections are rejected.
+- Verify: agent logs show TLS errors; no sensitive fallback to plaintext.
+
+6) gNOI server killed mid-upgrade
+- Setup: start a normal download, then kill/restart the gNOI server process on DUT during transfer/install.
+- Expected: agent detects stream error, marks session incomplete, and records state for resumption where supported.
+- Verify: session state is persisted; subsequent attempts either resume or fail with clear diagnostics.
+
+7) Disk corruption / I/O errors
+- Setup: emulate I/O errors or remove media during file write.
+- Expected: write fails; agent reports I/O error and cleans up partial files.
+- Verify: partial files removed; logs include I/O error details.
+
+Notes
+- Each scenario should include automation where possible (Ansible or test scripts) so PR-level tests can inject antagonists reliably.
+- Add per-scenario timeouts and retries to avoid false positives due to transient lab issues.
+
+### Test examples & code snippets
+Practical snippets and checks to include in test cases and automation.
+
+1) gNOI sanity checks (grpcurl)
+```sh
+# List services (plaintext test)
+grpcurl -plaintext DUT:50051 list
+# Describe the System service or method
+grpcurl -plaintext DUT:50051 describe System.SetPackage
+```
+
+2) Go helper: compute SHA‑256
+```go
+func computeSHA256(path string) ([]byte, error) {
+ f, err := os.Open(path); if err!=nil { return nil, err }
+ defer f.Close()
+ h := sha256.New()
+ if _, err := io.Copy(h, f); err!=nil { return nil, err }
+ return h.Sum(nil), nil
+}
+```
+
+3) Verification commands to include in tests
+```sh
+# Check file integrity on DUT
+sha256sum /tmp/SONiC.bin
+# Ensure gNOI service lists System and File services
+grpcurl -plaintext DUT:50051 list
+# Query version/state via gNMI (example)
+# gnmi-client get --target DUT --path /sonic/system/version
+```
+
+## Test Cases
```

Comment:
This section should move after `Physical SONiC Compatibility` Test Objective section.

First you could describe the happy-path (which is already here), and followed by negative cases for the  upgrade path.